### ISSUE TITLE
Tracking issue for RFC 2137: Support defining C-compatible variadic functions in Rust (c_variadic)

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#44930]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/44930)*

This is a tracking issue for the RFC "Support defining C-compatible variadic functions in Rust" (rust-lang/rfcs#2137).

**Steps:**

- [x] Implement the RFC (PR https://github.com/rust-lang/rust/pull/57760)
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
  - In particular, ensure we have all the UB documented. This includes the one for ABI compatibility: in particular, caller and callee must agree on whether the function is variadic and on the number of fixed arguments.
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

### Unresolved questions

- https://github.com/rust-lang/rust/issues/153646

#### passing `VaList` to `extern` functions

On linux calling `foo` will pass a `sysv64` `VaList` to an `extern "win64"` function, which might be confusing.

```
extern "win64" {
    fn foo(_: VaList);
}
```

### Resolved questions

**Miri support**

Has been added in https://github.com/rust-lang/rust/pull/150601.

**Continuing bikeshed on the `...` syntax**

There has been no discussion of a different syntax in recent years.

**When implementing this feature, we will need to determine whether the compiler can provide an appropriate lifetime that prevents a `VaList` from outliving its corresponding variadic function.**

The compiler internally determines that lifetime, to make sure that a `VaList<'_>` does not outlive the function that created it.

**Ensure that even when this gets stabilized for regular functions, it is still rejected on `const fn`.**

This is tested in https://github.com/rust-lang/rust/blob/main/tests/ui/parser/variadic-ffi-semantic-restrictions.rs and gives the following error:

```
error: functions cannot be both `const` and C-variadic
  --> $DIR/variadic-ffi-semantic-restrictions.rs:33:1
   |
LL | const unsafe extern "C" fn f4_1(x: isize, _: ...) {}
   | ^^^^^ `const` because of this             ^^^^^^ C-variadic because of this
```

We reject c-variadic `async` functions in a similar way.

**Issue https://github.com/rust-lang/rust/issues/125431**

- https://github.com/rust-lang/rust/issues/125431

The `VaList` definition introduced in https://github.com/rust-lang/rust/pull/141980 circumvents this issue.

**Issue https://github.com/rust-lang/rust/issues/61275**

- https://github.com/rust-lang/rust/issues/61275

This is an issue with c-variadic calls, not with c-variadic definitons.

**Issue https://github.com/rust-lang/rust/issues/141618**

- https://github.com/rust-lang/rust/issues/141618

We have some forwards-compatible ideas for how to potentially add the ABI to the type in the future. That is only required if we want to allow calling c-variadic functions with a non-system ABI. Because C only allows using the system ABI for c-variadic calls, that is unlikely to come up in practice. 

**What exactly are the ABI requirements around variadic calls? What are the ABI compatibility rules? Do we have to be concerned about LLVM target features affecting variadic ABI?**

Variadic calls using the system ABI do not impose any additional constraints. We already handle some ABI compatibility issues using target modifiers.

## Implementation history

**`va_arg` codegen**

- #56599
- #78126
- #141538
- #141622
- #141623
- #144549

**`VaList` and its layout**

- #49878
- #56396
- #57311
- #57760
- #58938
- #58985
- #59371
- #59625
- #62639
- #77784
- #126921
- #141980

**other**

- #126843
- #126927
- #141361
- #143619
- #144529
- #146165
- #146342
- #146434
- #146521
- #148770
- #149557

**c-variadic adjacent**

- #146551
- #146865

**Comment by plietar** [comment]
I'd like to work on this, I already have some prototype

**Comment by aturon** [comment]
Awesome @plietar! It'd probably be good to bring this up in [the "middle-end" compiler working group channel](https://gitter.im/rust-impl-period/WG-compiler-middle), which would also be a good place to get any help you might need.

**Comment by joshtriplett** [comment]
@plietar How goes the implementation? I remember you showing a mostly complete prototype on IRC.

**Comment by thedataking** [comment]
@plietar any news to share? This is a blocker for teams working on C to Rust transpilers, so this addition would be very welcome. (I'm part of one such team).

**Comment by plietar** [comment]
Hey,
Sorry I've been busy and then forgot about this. I'll get my prototype back in shape, hopefully by this weekend.

**Comment by harpocrates** [comment]
@plietar Any update on this? Do you have a WIP branch I can check out to try / fiddle with this?

**Comment by dlrobertson** [comment]
Looks like I'm a little late to the party :smile: ... sorry about that

A few questions:
1) How would functions that use a `va_list` multiple times work without the ability to explicitly use `va_start` and `va_end`? Are they expected to use copy? E.g. `execl` typically loops through the arguments to to get `argc`, creats a array `argv` of size `argc`, loops through the list again populating `argv`, and finally call `execv`.

2) The structure of a `va_list` varies greatly between the architectures. The intrinsic functions work with the architecture specific structure bitcast to an `i8*`, but AFAIK we'll still need to define the structure. Which architectures will be expected to be supported in the first iteration? Or am I mistaken that we'll have to define the structure?

@plietar if you don't have the time to work on this any more or if there is any way I could help out, I'd be more than happy to do so. I haven't worked on rustc much, but I'd be happy to help however I can with the implementation of this.

**Comment by nikomatsakis** [comment]
@dlrobertson 

Seems likely that @plietar doesn't have much time, though they can speak for themselves.

I've not really looked closely at what would be needed to implement this, but if you need any help, please ping me, or reach out on gitter/IRC.

**Comment by dlrobertson** [comment]
I've been working on this for the past two week and have

 - Added `va_list_kind` to the target specification which closely mirrors [clang TargetInfo::BuiltinVaListKind](http://clang.llvm.org/doxygen/classclang_1_1TargetInfo.html#a3502c200a07b649537ac95a3be7c8653)
 - Added `Type::va_list` which builds the correct structure based on the targets `va_list_kind`

I'm struggling a bit with understanding how to write something in `libcore` and link that to `Type::va_list` in trans. I'm currently attempting to add `#[lang = "va_list"]` so that I can check the `def.did` against the `lang_items().va_list_impl()` id. Does this seem correct?

**Comment by alexreg** [comment]
@dlrobertson That's super. Are you implementing the `...` syntax given that seems to be the consensus so far, and bikeshedding hasn't revealed a better one? (Unless I'm behind on things.)

**Comment by ehouarn-perret** [comment]
Am I wrong or variadic functions could be used to replace some macros like println / vec etc.?

**Comment by alexreg** [comment]
@ehouarn-perret Yes, they certainly could be. That may happen after they're implemented and stabilised... but anyway this feature is slightly different: it's about C-compatible variadic functions (`VaList`), not Rust-native variadics. The intention is to implement both in time.

**Comment by dlrobertson** [comment]
@ehouarn-perret as @alexreg mentioned this work is purely for C variadic functions, so it is only valid for `extern "C"` functions. I recently started researching variadic generics. I think that is what you're looking for.

**Comment by ehouarn-perret** [comment]
@alexreg
@dlrobertson
Thanks for pointing this out. True I was more looking at Rust native variadic support.
Sorry for the noise

**Comment by TheDan64** [comment]
@dlrobertson Do you have any update on this? If someone had the time, would it be possible to provide assistance in any way?

**Comment by dlrobertson** [comment]
@TheDan64 thanks for checking up on this, and there is definitely enough work to share! I'll use your prompt as a chance to give a general post on the status of my current work and the general state of things :smile:

## Defining "true" C variadic functions in Rust

 - [x] Update parsing to allow for implementing C variadic functions
 - [x] Spoof the `VaList` argument for Rust defined C variadic functions
 - [x] Inject `va_start`
 - [x] Inject `va_end`
 - [ ] Correctly call Rust defined C variadic functions in Rust

I plan to post a very WIP PR shortly.

**NB:** There are two know issues with the current codegen code:
 - The creation of the `va_list` allocates a `VaList`, which compiles to a `i64*`. It should 1) allocate a `VaListImpl` 2) get the pointer to the allocated `VaList` 3) call `va_start` on 2
 - The calculation of the argument name that `va_end` should be called on is incorrect.
 - The "spoofed" `VaList` causes calls to Rust defined C variadic functions from Rust to be incorrect.

## Other work unrelated to my current work

 - [ ] Fix Aarch64 support: https://github.com/rust-lang/rust/issues/56475
 - [ ] Attempt to implement `VaList::arg` in pure Rust: https://github.com/rust-lang/rust/issues/56489#issuecomment-444142735 also see [va_list-rs]
 - [ ] Cleanup the `cfg`s in `src/libcore/ffi.rs` with the use of `cfg_if`. Also see: https://github.com/rust-lang/rust/issues/57446
 - [ ] Test on various architectures and OSes
 - [ ] Improve documentation

If anyone is interested in working on these, please feel free to, and let me know how I can help!

[va_list-rs]: https://github.com/thepowersgang/va_list-rs

**Comment by dlrobertson** [comment]
\o/ https://github.com/rust-lang/rust/pull/57760 has been merged thanks to some awesome reviewing and help I received from @alexreg, @oli-obk, @matthewjasper, @varkor, and many others. The `c_variadic` feature now enables both `core::ffi::VaList` and Rust defined C-variadic functions. The work is still a long way from stable and there are still a few open issues, but I think *technically* the RFC has been implemented now. Could someone with the right privileges update the issue and tags?

**Comment by thedataking** [comment]
@dlrobertson We are working to use the `c_variadics` in the [C2Rust translator](https://github.com/immunant/c2rust). Things were going great until we took a closer look at the design of `VaList::copy`. It makes a lot of sense to ensure that the copy of the `va_list` can only be used inside the closure such that `copy` can call `va_end` before returning the result of the closure. However, this design makes syntax-directed translation from C really difficult, if not impossible, in some situations. See issue https://github.com/immunant/c2rust/issues/43 for more detail. 

On the other hand, exposing the `va_copy`  and `va_end` intrinsics would make a translation of the original C code straightforward. We'd be grateful for your thoughts on this. I think this is essentially what @joshtriplett  and @eddyb already suggested in these comments on the RFC:

- https://github.com/rust-lang/rfcs/pull/2137#issuecomment-328621331
- https://github.com/rust-lang/rfcs/pull/2137#issuecomment-328885252

**Comment by alexreg** [comment]
@thedataking `Drop` isn't getting used here for `va_end`. If you want those intrinsics to be exposed, we may be able to do that...

**Comment by dlrobertson** [comment]
> However, this design makes syntax-directed translation from C really difficult

It does take a bit more care and thought to convert code, and I'm not entirely sure how to do this automatically like c2rust is designed to do. It will take a lot to convince me that the API should be changed because I think the current API promotes a relatively safer use of `va_list`.

> On the other hand, exposing the va_copy and va_end intrinsics would make a translation of arbitrary C code straightforward.

Opinions changed as we worked on it (see [this comment](https://github.com/rust-lang/rust/pull/49878#discussion_r189441310)). Also, the intrinsics are not useful without a API change because we don't export the underlying implementation etc.

I'll post on the immunant issue and if there really is no other way, further discussion should be moved to the RFC PR as the people who worked on it would probably have some input.

**Comment by OlegTheCat** [comment]
Hi, guys. Sorry in advance, if it's the wrong place to ask. I'm currently trying to write a Rust function that accepts a vararg of C structs, but [`VaArgSafe`](https://doc.rust-lang.org/1.33.0/src/core/ffi.rs.html#144) trait, which is implemented only for a handful of types, doesn't allow me to do this. I believe that C language doesn't have such restrictions. 
Also, I've come across [this comment](https://github.com/rust-lang/rfcs/pull/2137#discussion_r136722075), which raises a similar point. 

Is it something that is going to be addressed in the future?

**Comment by alexreg** [comment]
@OlegTheCat Actually I believe the C standard does specify that this is undefined behaviour, though I don't have the relevant section at hand. (Incidentally I'm curious why f32 isn't supported.) I know @dlrobertson is busy currently, but hopefully he can clarify when he's back.

**Comment by dlrobertson** [comment]
@alexreg
> I'm curious why f32 isn't supported.

`float` is not allowed as stated in the C specification.

@OlegTheCat Once you get into arbitrary aggregate types things get much more complex. More testing and stabilization of the current implementation needs to be done to open `VaList::arg` up to more types. Are you working with an aggregate type or is there a primitive type that you're hitting this with?

**Comment by OlegTheCat** [comment]
> I believe the C standard does specify that this is undefined behaviour

@alexreg I've just had a glance at C standard and couldn't find any info regarding struct types and UB. As far as I can see, the only thing that is UB in varargs is when the type of the passed value and the type passed to `va_arg` don't match. Maybe I'm looking into the wrong section.

> I'm curious why f32 isn't supported.
> float is not allowed as stated in the C specification.

As far as I understand, the thing is that `f32` is being promoted to a `double` (`f64`) when passed as a vararg. Therefore there's no actual possibility to retrieve an `f32` value from a `va_list` because of the promotion.

@dlrobertson Yeah, I'm working with simple aggregate types and all of them have a `#[repr(C)]` tag. Something like this:

```
#[repr(C)]
struct Foo {
    x: i32
}

#[repr(C)]
struct Bar {
    foo: Foo
}


unsafe extern fn vararg_test(n: usize, mut args: ...) {
    ...
    args.arg::<Bar>(); //this cannot be compiled
    ...
}
```

**Comment by alexreg** [comment]
> As far as I understand, the thing is that `f32` is being promoted to a `double` (`f64`) when passed as a vararg. Therefore there's no actual possibility to retrieve an `f32` value from a `va_list` because of the promotion.

This is what I understood too. In that case, there's no possibility of retrieving anything other than a C `int`, `unsigned int`, or `double`... so why the impls for the other types?

**Comment by joshtriplett** [comment]
On May 28, 2019 2:34:10 PM PDT, Alexander Regueiro <notifications@github.com> wrote:
>> As far as I understand, the thing is that `f32` is being promoted to
>a `double` (`f64`) when passed as a vararg. Therefore there's no actual
>possibility to retrieve an `f32` value from a `va_list` because of the
>promotion.
>
>This is what I understood too. In that case, there's no possibility of
>retrieving anything other than a C `int`, `unsigned int`, or
>`double`... so why the impls for the other types?

Precisely because those impls are supposed to handle the promotion behavior correctly.

**Comment by alexreg** [comment]
Ah, I see, fair enough then.

**Comment by eddyb** [comment]
One thing we need to not lose track of is: `VaList` shouldn't be leaking into the `ty::FnSig` of functions that don't pass `VaList` as a regular argument (but rather they're variadic and use `VaList` internally).

I think the way we should handle this is:
* AST: whatever matches the syntax best
  * e.g. `ast::TyKind::CVarArg` is probably fine
* HIR/MIR: mark the signature as C-variadic instead of having a type for the last argument
  * the argument list inside the body would be one longer than the signature, *but this is fine*
  * anything dealing with those arguments lists can detect this from the signature being C-variadic
  * `rustc_{typeck,mir}` can give that argument pattern the type `VaList`, as it's a lang item
  * codegen can special-case the argument (see also: the `spread_arg` field of `mir::Body`)

From the outside, a Rust `fn` definition that's C-variadic would look no different than:
```rust
extern {
    fn foo(a: A, b: B, c: C, ...);
}
```

**Comment by DragoonAethis** [comment]
Is there any interest towards implementing this RFC for non-native extern ABIs? I'm primarily interested in "win64" ABIs, as that's what is used under UEFI (there are some platform APIs that have variadic args and they're required to implement certain driver functionality).

**Comment by dlrobertson** [comment]
@DragoonAethis Do you have a link to a spec or good implementation? I found some basic info, but nothing solid. This could be done. I don't _think_ `win64` is covered by LLVM, so I think we'd need to implement `va_arg` etc.

**Comment by DragoonAethis** [comment]
It's documented by Microsoft [here](https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019), although quick googling also suggests LLVM doesn't support this. (I'm also not an expert, so I'd be happy to be wrong :)

It looks like varargs on Windows are just loaded as next arguments with some exceptions like "float args are always converted to doubles". There's a test suite from .NET Core available [here](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/jit64/mcc/interop) that tries to check quite a few cases for correctness, so one could browse these to see if there are any special gotchas in there.

**Comment by eddyb** [comment]
> with some exceptions like "float args are always converted to doubles"

Pretty sure that's enshrined in the C standard (the promotion happens in the call, *before* ABI).

**Comment by dlrobertson** [comment]
> Pretty sure that's enshrined in the C standard (the promotion happens in the call, before ABI).

Yeah floats are not allowed, so that would have to happen before.

**Comment by Stargateur** [comment]
Variadic in C are maybe the definition of implemented behavior, well maybe bitfield are...

Relevant section of C11 (I don't have C17, I think its change):

> [6.5.2.2.6](https://port70.net/~nsz/c/c11/n1570.html#6.5.2.2p6) If the expression that denotes the called function has a type that does not include a prototype, the integer promotions are performed on each argument, and arguments that have type float are promoted to double. These are called the default argument promotions. If the number of arguments does not equal the number of parameters, the behavior is undefined. If the function is defined with a type that includes a prototype, and either the prototype ends with an ellipsis (, ...) or the types of the arguments after promotion are not compatible with the types of the parameters, the behavior is undefined. If the function is defined with a type that does not include a prototype, and the types of the arguments after promotion are not compatible with those of the parameters after promotion, the behavior is undefined, except for the following cases:
> 
> - one promoted type is a signed integer type, the other promoted type is the corresponding unsigned integer type, and the value is representable in both types;
> - both types are pointers to qualified or unqualified versions of a character type or void.
>
> 7 If the expression that denotes the called function has a type that does include a prototype, the arguments are implicitly converted, as if by assignment, to the types of the corresponding parameters, taking the type of each parameter to be the unqualified version of its declared type. The ellipsis notation in a function prototype declarator causes argument type conversion to stop after the last declared parameter. The default argument promotions are performed on trailing arguments.
> 
> 8 No other conversions are performed implicitly; in particular, the number and types of arguments are not compared with those of the parameters in a function definition that does not include a function prototype declarator.

Have fun for the next one:

> [6.3.1.11](https://port70.net/~nsz/c/c11/n1570.html#6.3.1.1p1) Every integer type has an integer conversion rank defined as follows:
> 
> - No two signed integer types shall have the same rank, even if they have the same representation.
> - The rank of a signed integer type shall be greater than the rank of any signed integer type with less precision.
> - The rank of long long int shall be greater than the rank of long int, which shall be greater than the rank of int, which shall be greater than the rank of short int, which shall be greater than the rank of signed char.
> - The rank of any unsigned integer type shall equal the rank of the corresponding signed integer type, if any.
> - The rank of any standard integer type shall be greater than the rank of any extended integer type with the same width.
> - The rank of char shall equal the rank of signed char and unsigned char.
> - The rank of _Bool shall be less than the rank of all other standard integer types.
> - The rank of any enumerated type shall equal the rank of the compatible integer type (see 6.7.2.2).
> - The rank of any extended signed integer type relative to another extended signed integer type with the same precision is implementation-defined, but still subject to the other rules for determining the integer conversion rank.
> - For all integer types T1, T2, and T3, if T1 has greater rank than T2 and T2 has greater rank than T3, then T1 has greater rank than T3.
> 
> 2 The following may be used in an expression wherever an int or unsigned int may be used:
> - An object or expression with an integer type (other than int or unsigned int) whose integer conversion rank is less than or equal to the rank of int and unsigned int.
> - A bit-field of type _Bool, int, signed int, or unsigned int.
> 
> If an int can represent all values of the original type (as restricted by the width, for a bit-field), the value is converted to an int; otherwise, it is converted to an unsigned int. These are called the integer promotions.58) All other types are unchanged by the integer promotions.
> 3 The integer promotions preserve value including sign. As discussed earlier, whether a ''plain'' char is treated as signed is implementation-defined.

The [stdarg.h](https://port70.net/~nsz/c/c11/n1570.html#7.16.1) definition.

So, AFAIK, "the integer promotions are performed on each argument" say that a structure type is also rule by the integer promotion so a structure that only contain a char should be promote to int in size, enum too and union too. The only exception is floating number.

But, I never seen a (serious) code that push a structure in a variadic list.

**Comment by sarvi** [comment]
Just FYI. One problem I ran into is the rust macro cannot deal with "..."
If I wanted to create a rust macro for hook 
```
hook! {
    unsafe fn prinf(format: *const c_char, args:...) -> c_int => my_prrintf {
        if let Ok(path) = std::str::from_utf8(std::ffi::CStr::from_ptr(path).to_bytes()) {
            println!("printf(\"{}\")", format);
        } else {
            println!("printf(...)");
    }
}
```
The goal of the macro is to expand to a larger rust function that also takes c_variadic.

if hook macro was defined as follows, it cant capture "..." as a type and errors out.
```
unsafe fn $real_fn:ident ( $($v:ident : $t:ty),* ) -> $r:ty => $hook_fn:ident $body:block) => {
.....
```

**Comment by sarvi** [comment]
One more issue. C_variadics only works as functions not as methods.
The following function works
```
#![feature(c_variadic)]
extern crate libc;
use libc::{c_char,c_int};
#[no_mangle]
pub unsafe extern "C" fn printf(_format: *const c_char, mut args: ...) -> c_int  {
    10
}
```

Done as a method, it fails
```
#![feature(c_variadic)]

extern crate libc;

// #[macro_use]
extern crate redhook;

use libc::{c_char,c_int};


#[allow(non_camel_case_types)]
pub struct printf {__private_field: ()}
#[allow(non_upper_case_globals)]
static printf: printf = printf {__private_field: ()};


impl printf {
    #[no_mangle]
    pub unsafe extern "C" fn printf(_format: *const c_char, mut args: ...) -> c_int  {
        10
    }
}
```

The non c_variadic version of the above code compiles fine
```
#![feature(c_variadic)]

extern crate libc;

// #[macro_use]
extern crate redhook;

use libc::{c_char,c_int};


#[allow(non_camel_case_types)]
pub struct printf {__private_field: ()}
#[allow(non_upper_case_globals)]
static printf: printf = printf {__private_field: ()};


impl printf {
    #[no_mangle]
    pub unsafe extern "C" fn printf(_format: *const c_char) -> c_int  {
        10
    }
}
```

**Comment by jethrogb** [comment]
That's not a method, but an associated function. I do think associated functions should support this feature though.

**Comment by dlrobertson** [comment]
> That's not a method, but an associated function. I do think associated functions should support this feature though.

Yeah, associated functions do not support this feature at the moment. Associated functions were not in the original RFC, but I agree that they should be supported. Will post a PR in a sec.

**Comment by joshtriplett** [comment]
On Sat, Jul 25, 2020 at 12:52:51PM -0700, Dan Robertson wrote:
> > That's not a method, but an associated function. I do think associated functions should support this feature though.
> 
> Yeah, associated functions do not support this feature at the moment. Associated functions were not in the original RFC, but I agree that they should be supported. Will post a PR in a sec.

Thanks!

**Comment by nikomatsakis** [comment]
Apart from #74765, what is the status of this work? Is this something we could move towards stabilization? Maybe someone would be interested in driving that? Nominating for @rust-lang/lang meeting.

**Comment by joshtriplett** [comment]
We discussed this in the @rust-lang/lang meeting today. We feel like this may be ready for a stabilization report and stabilization PR. Associated function support *just* went in, and it doesn't feel like it would be worth the trouble to split the feature gate for that, so we could instead wait one release for that to bake.

**Comment by dlrobertson** [comment]
https://github.com/rust-lang/rust/pull/73655#event-3576995068 adds aarch64 support, but I have not been able to test on many architectures. Are there certain architectures/OSes that we'd like to have tested better before stabilization?

**Comment by joshtriplett** [comment]
@dlrobertson You might consider porting some of the more complex tests from libffi (`testsuite/va_*.c`) to Rust. Those test lots of corner cases. If we have those, and they pass on a new architecture, and that architecture's support goes in at the beginning of a new development cycle (right after beta branches), I wouldn't have any concerns.

**Comment by nikomatsakis** [comment]
By the way, I would like us to get in the habit of posting IntoRust blog posts if we'd like folks to experiment -- maybe somebody wants to write a IntoRust blog post highlighting that we plan to stabilize this feature and encouraging folks  to tinker with it?

**Comment by dlrobertson** [comment]
Midway through implementation I talked at a local meetup about it. Would [these slides](http://dlrobertson.com/slides/va-list-12-13-2018.html#/) be (at least partially) salvageable.

**Comment by Grinkers** [comment]
I, like probably many others that rely heavily on c, would really like to see this stabilized.

As a recommended next step, I started porting over some of libffi's more complex tests, but quickly ran into a limitation with #[repr(C)] struct's.  As of right now, we're limited to only these types, even though C can pass structures.
https://github.com/rust-lang/rust/blob/a3ed564c130ec3f19e933a9ea31faca5a717ce91/library/core/src/ffi.rs#L309

I have some availability to work on this, but am rather unfamiliar with contributing to Rust. So I thought I'd at least bring up the issue here for discussion/advice.

**Comment by joshtriplett** [comment]
@Grinkers I'd like to see support for passing structs by value, but for now, could you port over the tests that *don't* depend on that functionality, or stub out the struct passing?

**Comment by Grinkers** [comment]
Sure, just to be clear, I assume you mean the va_*.c https://github.com/libffi/libffi/tree/master/testsuite/libffi.call which are mostly the same and testing structs. Without the structs it's mostly just testing various sized c types, which is still very good to test on various architectures, but it seems like we cover quite a few.

Is there any automated tests different architectures or is that work that needs to be done?

**Comment by jonas-schievink** [comment]
Currently it looks like `...` is syntactically allowed *anywhere* in the parameter list. Is that intentional, or should it be restricted to the last parameter only?

**Comment by jonas-schievink** [comment]
Ah I guess that ship has sailed, since this is accepted on stable:

```rust
extern {
    fn f(#[cfg(NEVER)] ..., named: u8, ...);
}
```

**Comment by varkor** [comment]
@jonas-schievink: this seems like a bug according to the RFC.

> This must appear as the last argument of the function, and the function must have at least one argument before it.

I don't imagine this behaviour is being relied upon, so we can probably fix it now.

**Comment by nikomatsakis** [comment]
I agree that is a bug, possibly even a regression, though who knows when it may have been introduced.

**Comment by cuviper** [comment]
Note that s390x hits an `LLVM ERROR` in the test for this feature -- #84628.

**Comment by Aaron1011** [comment]
Is anything blocking the stabilization of this? s390x is a tier-two target, so could we just emit a warning/error on LLVM versions (currently all of them) where https://github.com/rust-lang/rust/issues/84628 occurs?

**Comment by dlrobertson** [comment]
s390x should fall through to LLVM's `va_arg` instruction, so there might be an issue there. I can't remember off the top of my head, but I think s390x uses the void pointer variant of a `va_list`, so I wonder if using our `emit_ptr_va_arg` like we do for `x86` and windows would fix some of the issues seen on s390x. I haven't tested on s390x, so I wouldn't know for sure though.

**Comment by cuviper** [comment]
I don't know these details well, but I can volunteer to test things on s390x hardware if needed.

**Comment by cuviper** [comment]
FWIW, here is clang's `SystemZABIInfo::EmitVAArg`:

https://github.com/llvm/llvm-project/blob/9ad17fe0debb0b48798aaba6d07b65136fcf0664/clang/lib/CodeGen/TargetInfo.cpp#L7467-L7468

**Comment by dlrobertson** [comment]
Thanks for the link! There is definitely something wrong with our implementation. We're emitting a void pointer like va_list but it looks like a complex structure is expected. We'll need to change the intrinsic type `VaListImpl` for the architecture. It would be interesting to see if just changing the structure for s390x fixes this or if we do indeed need to implement a custom va_arg as well.

**Comment by cuviper** [comment]
Let's take the s390x discussion back to #84628.

**Comment by HTG-YT** [comment]
I see a potential use case for `efiapi` convention: 
![image](https://user-images.githubusercontent.com/39023054/151487526-19bd320f-11b6-43cb-af82-a2b4475372b6.png)

Example implementation:
```rust
pub type EFI_INSTALL_MULTIPLE_PROTOCOL_INTERFACES = extern "efiapi" fn(
    Handle: *mut VOID,
    ...
) -> EFI_STATUS;
```

Is this out of the scope of this RFC, or should a new RFC be opened to extend this feature?

**Comment by jethrogb** [comment]
Isn't EFIAPI not just one of the existing Microsoft calling conventions?

**Comment by HTG-YT** [comment]
Probably, but I was thinking about would `extern "C"` work here - as it stands, I am working with EFI, *not* C.......?

Please do correct  and clarify if I am mistaken. Corrections are appreciated. 👍

**Comment by GabrielMajeri** [comment]
@jethrogb Not exactly. See also this issue with the `uefi-rs` library: https://github.com/rust-osdev/uefi-rs/issues/302

The basic idea is: for the GNU C/C++ compilers, you can choose a calling convention (`extern "C"` or just no extern), and you can also choose an ABI using [a custom function attribute](https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html) (`ms_abi` vs `sysv_abi`).

Rust doesn't have this distinction. There's no `extern "C" with ms_abi` or `extern "C" with gnu_abi`; because of this, we're kind of forced to either use `extern "C"` everywhere (and the ABI cannot be selected, it depends on whatever settings the Rust compiler has for that target; which is risky) or use `extern "efiapi"` (which will enforce the MS ABI, but doesn't currently work with variadics, as mentioned [in the issue](https://github.com/rust-osdev/uefi-rs/issues/302)).

**Comment by Stargateur** [comment]
> I am working with EFI, not C.......?

C standard don't state any ABI requirement. Rust `extern "C"` is "use the classic C ABI for this target used by most C compiler", it's not guarantee to work all the time.

**Comment by dlrobertson** [comment]
Perhaps we should also allow `sym::efi_api` in [`check_variadic_type`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_passes/src/ast_validation.rs#L577)? `extern efi_api` didn't exist when I implemented it, which is the only reason I didn't add it in the original implementation. I don't immediately see a reason we shouldn't allow `extern "efiapi"`. I could take a crack at an implementation attempt if this is a reasonably common use case. Issues that I think we'll hit:

  1) I highly doubt we'd emit the correct structure for the `VaListImpl`. `clang` emits a void pointer va list, but I'm pretty sure for x86_64 and aarch64 we'd emit the structure used by both. I don't think this would be too hard to fix. We could just check `target_os`, which should be `"uefi"`. We don't currently handle that properly, which is a bug.
  2) `clang` uses `emitVoidPointerVaArg` in this case, so I'd assume we'd need to do the same and add a case to our llvm backend.

Questions:

 - I'm not up to speed with `extern "efiapi"`. Would there be a case where "efiapi" was used but `target_os != "uefi"`? We figure out the right va_list structure and code to emit based on the `target_os` and `target_arch`, so if this were to work we'd need `target_os == "uefi"` any time we expect to use that ABI.
 - Is there a test case for this? I have minimal uefi experience.

**Comment by HTG-YT** [comment]
> I could take a crack at an implementation attempt if this is a reasonably common use case.

I am not sure whether it would be a *common* use case for *everything*; but it would indeed be a huge demand for support in low-level development, specifically when working with the UEFI specification and even OSDev with Rust in general.

**Comment by GabrielMajeri** [comment]
> Would there be a case where "efiapi" was used but target_os != "uefi"? We figure out the right va_list structure and code to emit based on the target_os and target_arch, so to get this to work, I think we'll need target_os == "uefi" any time we expect to use that ABI.

Yes, unfortunately... Besides directly targeting UEFI and using the `extern "efiapi"` functions with the native MS ABI on UEFI targets, it's also possible that your Rust app is actually an OS kernel (targeting ELF with the GNU ABI for example), but you want to call some firmware functions which are declared with `extern "efiapi"` (MS ABI calling convention).

**Proposed solution**: I'd think it's enough to force "ms_abi" on all `extern "efiapi"` functions (since I'm pretty sure _that's_ how that calling convention is defined by the spec - unlike regular `extern "C"` functions which could be `ms_abi` or `gnu_abi` dependening on the target OS) and add support for C variadics for them, just like for regular `extern "C"` functions.

**Comment by dlrobertson** [comment]
:+1: I think this enough info to get started on a implementation. To start, inspecting the LLVM IR emitted by simple standalone examples should be enough, but to move from WIP to something merge-able, I'd like a real test case. For that I might need some pointers and/or some help. Is there a commonly used IRC channel or chat mechanism for the rust-osdev group? If not, I may move discussions about test cases to https://github.com/rust-osdev/uefi-rs/issues/302.

**Comment by dlrobertson** [comment]
Did a bit of work on this and `x86_64-unknown-uefi` emits the correct LLVM now, but I've realized we hit issues with specific cases for `extern "efiapi"`. We currently emit one `VaListImpl` as a language item in `library/core/ffi.rs` based on the OS and architecture. This causes issues for the following example:

```
#![no_std]
#![feature(c_variadic)]
#![feature(abi_efiapi)]

pub unsafe extern "efiapi" fn my_example_fn(int: u64, mut args: ...) -> u64 {
    if int > 0 {
        args.arg::<u64>()
    } else {
        0
    }
}
```

If you compile with `rustc --target=x86_64-unknown-linux-gnu --crate-type=lib --emit=llvm-ir <file>`, you can find something like the following

```
%"core::ffi::VaListImpl" = type { <va list structure here> }
...
define win64cc i64 ...
  ...
  ; this isn't right. inside this function the VaListImpl should be a void pointer
  %args = alloca %"core::ffi::VaListImpl", align 8
  %2 = bitcast %"core::ffi::VaListImpl"* %args to i8*
  call void @llvm.va_start(i8* %2)
  ...
  <emit_ptr_va_arg code here>
```

So currently defining new C variadic functions with `extern "efiapi"` doesn't really work when the target doesn't match. I think I have a solution though. The fix might not be the easiest, but I think what we can do is create a `MsVaListImpl` that is a type that always equals the void pointer variant. Then if the function is `extern "efiapi"` use that language item instead of the VaListImpl. The downside is that we're adding a language item and potential complexity. Any thoughts, questions, or input before I make an attempt at this?

**Comment by joshtriplett** [comment]
Nominating this for the next @rust-lang/lang meeting, so that we can make a decision about what targets need to be supported before we stabilize this.

In the past, for things like `i128`/`u128`, we added and stabilized a mechanism without waiting for every last target to implement it. I think we need to do something similar here, rather than waiting for *every* target to completely implement `...`. We could, for instance, mark `VaListImpl` with `cfg` so that code using it on an unsupported target doesn't compile.

Separately, we also need to decide if we can forwards-compatibly handle variable-argument functions for different ABIs.

**Comment by joshtriplett** [comment]
@dlrobertson It'd help to have a rough summary of the current status. Which targets does it work completely on, without any holes as far as we know? Which targets does it work *partially* on, but still needs fixes?

Also, is there a sketch for how to handle multiple calling conventions on the same target, in a forwards-compatible way? We can stabilize this on a target without supporting any non-default calling conventions on the target, but we need to feel confident that we can expand to support non-default calling conventions in the future without breaking changes. If we need to change something before stabilization to leave ourselves room for evolution here, we should. (For instance, does VaList need to have an opaque template parameter for ABI?)

**Comment by dlrobertson** [comment]
@joshtriplett

> It'd help to have a rough summary of the current status. Which targets does it work completely on, without any holes as far as we know? Which targets does it work partially on, but still needs fixes?

AFAIK all tier 1 targets are supported for `{i,u}{8,16,32,64,size}` and `f64`. I have not tested targets beyond that enough to claim reasonable support, and as a result there are bugs to be found there. The targets I know there are issues with are `target_os = "uefi"` and `target_arch = "s390x"`. FWIW a good way to test if we can support a target is to run the `c-link-to-rust-va-list-fn` test in `run-make-fulldeps`. If that test passes, I'm comfortable enough with our support of the target to maintain it.

> Also, is there a sketch for how to handle multiple calling conventions on the same target, in a forwards-compatible way?

https://github.com/rust-lang/rust/issues/44930#issuecomment-1027941470 adding a language item that is always the pointer variant is the only way I can currently see us supporting the target + the MS ABI.

> We can stabilize this on a target without supporting any non-default calling conventions on the target, but we need to feel confident that we can expand to support non-default calling conventions in the future without breaking changes.

:+1: I agree

> If we need to change something before stabilization to leave ourselves room for evolution here, we should. (For instance, does VaList need to have an opaque template parameter for ABI?)

I'm not entirely sure I follow here. I don't know how we could get the appropriate layout in codegen from this, but I'm likely missing something. Since we really need two implementations of variadics for a given libcore build, I think the easiest would be to add another lang item, and based on the function ABI inject the correct type. That being said, I also would like to avoid adding another lang item if possible. So if using an opaque template parameter would work, I think it would be preferred.

Are there any other language items that have a variable layout depending on the ABI? Another implementation that could serve as a guide would be hugely helpful.

**Comment by nikomatsakis** [comment]
I agree with @joshtriplett that we should stabilize this on a per-target basis.

**Comment by joshtriplett** [comment]
Tagging as ready-to-stabilize based on previous lang discussion.

The author mentioned wanting to add some more tests. Assuming those tests pass, this is ready for stabilization; if any fail, this should move to impl-incomplete until they pass.

**Comment by rayanmargham** [comment]
Just commenting to let everyone know that this hasn't been completed just yet

**Comment by jondo2010** [comment]
Is there any movement on this? Last comment is from more than 12 months ago.

**Comment by dlrobertson** [comment]
> Is there any movement on this? Last comment is from more than 12 months ago.

I haven't had much free time to work on this. Is there a particular open issue that is part of stabilizing Rust defined C-variadics you're experiencing issues with?

**Comment by dlrobertson** [comment]
> I just tested my limited use case, (tier1 only; linux and windows), without issues.
> 
> https://github.com/rust-lang/rust/tree/master/tests/run-make/c-link-to-rust-va-list-fn Are there more tests that are required before moving to stabilization?

I'd like to add support for defining functions with an EFI API within a x86_64, aarch64, or powerpc target, but not sure if this is needed to be completed before stabilization.

**Comment by Jules-Bertholet** [comment]
Bikeshed: the ellipsis in `mut args: ...` has no precedent anywhere else in the language. The syntax could be something more reminiscent of pattern matching, like `mut args @ ..`.

**Comment by Stargateur** [comment]
> Bikeshed: the ellipsis in `mut args: ...` has no precedent anywhere else in the language. The syntax could be something more reminiscent of pattern matching, like `mut args @ ..`.

I'm pretty sure that would be a bad thing, pattern matching is already a thing in function argument; you can deconstruct a tuple for example. And something like `fn foo(args @ ..: &[u8]) { }` is already illegal. The proposition to add a special type `...` make more sense.

**Comment by Jules-Bertholet** [comment]
> pattern matching is already a thing in function argument

Exactly, so we should consider following that precedent.

> vaarg is not a pattern it's a type.

When you are calling the variadic function, you provide a comma-separated set of values—which is exactly what `..` matches in patterns.

**Comment by dlrobertson** [comment]
> @dlrobertson I may be misunderstanding you, but I think variadics already work fine with efiapi? Probably from #97971. I've been using `c_variadic` with `x86_64-unknown-uefi` and `i686-unknown-uefi` without issue for some time.

Great point! This isn't quite what I'm getting at though. See [this rust example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=77fb45833eed34f73105d45f0aa36a65) and note that the `VaList` structure used by the `"efiapi"` function is actually the ABI of the host (in this case amd64). AFAIK the ABI of the `VaList` in the example should be the void pointer variant. I'm not entirely sure how often such a case would exist in the wild.

Also see [this comment](https://github.com/rust-lang/rust/issues/44930#issuecomment-1027941470) (It's a little outdated, but I think there are parts of it that are still somewhat relevant).

**Comment by nicholasbishop** [comment]
I see, thanks for the clarification!

**Comment by Grinkers** [comment]
> > @dlrobertson I may be misunderstanding you, but I think variadics already work fine with efiapi? Probably from #97971. I've been using `c_variadic` with `x86_64-unknown-uefi` and `i686-unknown-uefi` without issue for some time.
> 
> Great point! This isn't quite what I'm getting at though. See [this rust example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=77fb45833eed34f73105d45f0aa36a65) and note that the `VaList` structure used by the `"efiapi"` function is actually the ABI of the host (in this case amd64). AFAIK the ABI of the `VaList` in the example should be the void pointer variant. I'm not entirely sure how often such a case would exist in the wild.
> 
> Also see [this comment](https://github.com/rust-lang/rust/issues/44930#issuecomment-1027941470) (It's a little outdated, but I think there are parts of it that are still somewhat relevant).

There was previous discussion about stabilization on a per-target basis. Would gating `"efiapi"` for now be sufficient (while still leaving room for the future)?

I feel it's a shame to have this stuck in limbo for something that may not have a use case. Especially when this is a blocker for rust adoption in many cases on tier 1 targets. Or, if this target issue is problematic, I think it would be nice to have this in the Steps on the first post.

**Comment by tgross35** [comment]
Is there a way to make `VaArgSafe` visible in docs so it is easy to find allowed types? I am not sure if this is possible for sealed traits, but it would be nice if it was quick to find this information

**Comment by dlrobertson** [comment]
> Is there a way to make `VaArgSafe` visible in docs so it is easy to find allowed types? I am not sure if this is possible for sealed traits, but it would be nice if it was quick to find this information

Would it be reasonable add the documentation to the [VaList](https://doc.rust-lang.org/stable/std/ffi/struct.VaList.html). We currently only state:

> A wrapper for a `va_list`.

Perhaps we should add more info about the types allowed.

**Comment by nicholasbishop** [comment]
Following up on the concern in https://github.com/rust-lang/rust/issues/44930#issuecomment-1551747344, I think that may not matter for stabilizing this because it's gated by a different feature: [`extended_varargs_abi_support`](https://github.com/rust-lang/rust/issues/100189).

To summarize my understanding of the concern, the problem occurs when you have an `extern "efiapi"` variadic function, _and_ you compile for a target that uses a different variadic ABI (e.g. `x86_64-unknown-linux-gnu`). See https://github.com/rust-lang/rust/issues/44930#issuecomment-1024849683 for an explanation of when that would come up.

But that case cannot currently compile: ``error: only foreign or `unsafe extern "C"` functions may be C-variadic``. There are a number of tests for this as well.

https://github.com/rust-lang/rust/issues/100189 tracks the `extended_varargs_abi_support` feature which would enable ABIs other than `"C"`. And personally that is something I would like to see in the future, but I don't think it needs to block stabilization of `c_variadic`.

@dlrobertson in case I've misunderstood anything above :)
@Soveu FYI since you have https://github.com/rust-lang/rust/pull/116161 open to stabilize `extended_varargs_abi_support`

**Comment by nicholasbishop** [comment]
Of course, now that I've typed that, I see I did miss something :) The example in https://github.com/rust-lang/rust/issues/44930#issuecomment-1027941470 won't compile because it uses `extern "efiapi"` with `...`, but the rust-playground example in https://github.com/rust-lang/rust/issues/44930#issuecomment-1551747344 _does_ compile because it uses `VaImpl` directly, and produces wrong code depending on the host target.

**Comment by dlrobertson** [comment]
> Of course, now that I've typed that, I see I did miss something :) The example in [#44930 (comment)](https://github.com/rust-lang/rust/issues/44930#issuecomment-1027941470) won't compile because it uses `extern "efiapi"` with `...`, but the rust-playground example in [#44930 (comment)](https://github.com/rust-lang/rust/issues/44930#issuecomment-1551747344) _does_ compile because it uses `VaList` directly, and produces wrong code depending on the host target.

Yeah, this is tricky to fix... Honestly, this case is probably trickier to fix than the c_variadic case. I'm uncertain how much of an edge case this is, so I don't know if this is a blocker or not.

**Comment by charmoniumQ** [comment]
Is there a way to "forward" the variadic arguments to another variadic function, supposing I know the size in bytes of the arguments?

Use-case: Implementing library interposition (aka LD_PRELOAD wrapper) around a C interface with variadic functions. For a (rough) example:

```rust
pub unsafe extern "C" fn printf(fmt: *const libc::c_char, mut args: ...) -> usize {
    let real_printf = dlsym(RTLD_NEXT, "printf");
    let size = /*
        compute size of args based on %-codes in fmt until reaching null byte
    */;
    real_printf(fmt, *args) /* How to forward args to real_printf? */
}
```

**Comment by pervognsen** [comment]
> ```rust
> ...
> ```

I need to do this too. Between reading the RFC (which appears outdated since there's now the split between VaList and VaListImpl) and looking at the scant API docs, my example code here seems to work: https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=a536485d2935d6138b6bd8425d5f43a0. However, it would be nice to have confirmation from someone who understands the intended semantics better.

**Comment by pervognsen** [comment]
> Is there a way to "forward" the variadic arguments to another variadic function, supposing I know the size in bytes of the arguments?
> 
> Use-case: Implementing library interposition (aka LD_PRELOAD wrapper) around a C interface with variadic functions. For a (rough) example:
> 
> ```rust
> pub unsafe extern "C" fn printf(fmt: *const libc::c_char, mut args: ...) -> usize {
>     let real_printf = dlsym(RTLD_NEXT, "printf");
>     let size = /*
>         compute size of args based on %-codes in fmt until reaching null byte
>     */;
>     real_printf(fmt, *args) /* How to forward args to real_printf? */
> }
> ```

I need to do this too. Between reading the RFC (which appears outdated since the current implementation now has a split between VaList and VaListImpl) and looking at the scant API docs, my example code here seems to work: https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=a536485d2935d6138b6bd8425d5f43a0. However, it would be nice to have confirmation from someone who understands the intended semantics.

**Comment by charmoniumQ** [comment]
We're talking about slightly different things.

<table>
<tr>
 <td>
 <td>Function we are defining
 <td>Function we are calling
<tr>
 <td>My request
 <td>variadic (e.g., `printf_wrapper`)
 <td>variadic (e.g., `printf`)
<tr>
 <td>Your request
 <td>variadic (e.g., `printf_wrapper`)
 <td>not variadic (e.g., `vprintf`)
</table>

In GCC, forwarding variadic functions to a variadic function can be done using [`__builtin_apply`](https://gcc.gnu.org/onlinedocs/gcc/Constructing-Calls.html). I don't think there is a Rust equivalent of this.

**Comment by pervognsen** [comment]
> We're talking about slightly different things.
> 	Function we are defining 	Function we are calling
> My request 	variadic (e.g., `printf_wrapper`) 	variadic (e.g., `printf`)
> Your request 	variadic (e.g., `printf_wrapper`) 	not variadic (e.g., `vprintf`)
> 
> In GCC, forwarding variadic functions to a variadic function can be done using [`__builtin_apply`](https://gcc.gnu.org/onlinedocs/gcc/Constructing-Calls.html). I don't think there is a Rust equivalent of this.

Ah. It doesn't even look like clang supports __builtin_apply, which does not make me hopeful there would be a 1:1 equivalent in LLVM. Since the '...' in 'args: ...' currently stands for a platform/ABI-specific opaque VaListImpl, you can get 'naive forwarding' to pass Rust's type checks and it misleadingly passes basic tests as long as you stay within the ABI argument limit that stays in registers, but it's revealed as incorrect when you exercise a longer argument list: https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=3f7e586b0d35e50626cf290097731d6a. I'm assuming you got this, so I'm just leaving this here in case anyone else comes across this thread.

**Comment by bjorn3** [comment]
Something to note in the unresolved questions:

libcore currently hard codes the layout of va_list that llvm uses for each individual target. This probably won't work with other backends. How should we handle the target and backend dependent `VaListImpl` layout?

**Comment by lolbinarycat** [comment]
> Bikeshed: the ellipsis in `mut args: ...` has no precedent anywhere else in the language. The syntax could be something more reminiscent of pattern matching, like `mut args @ ..`.

untrue, it is already used when declaring extern variadic functions.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=724901e4535cd7c5981de0eff52a978f

**Comment by lolbinarycat** [comment]
presumably the syntax for defining variadic functions would also be a special case.

regardless, my point is that it does have precedent, and indeed is already associated with variadics.

**Comment by bjorn3** [comment]
For C-SKY the abi specification doesn't seem to dictate any specific layout for `va_list`. It suggests a possible implementation for `va_start` and `va_arg` in section 2.2.4, but does not say that this is the only valid implementation. And the Hexagon abi specification doesn't even mention `va_list`, `va_start` or `va_arg`. It only explains the calling convention in section 5.2.

You are right that many other abi specifications do explicitly mention the layout. I wasn't aware of this.

References:
C-SKY: https://github.com/c-sky/csky-doc/blob/master/C-SKY_V2_CPU_Applications_Binary_Interface_Standards_Manual.pdf
Hexagon: https://lists.llvm.org/pipermail/llvm-dev/attachments/20190916/21516a52/attachment-0001.pdf

**Comment by workingjubilee** [comment]
> On May 28, 2019 2:34:10 PM PDT, Alexander Regueiro ***@***.***> wrote: > As far as I understand, the thing is that `f32` is being promoted to a `double` (`f64`) when passed as a vararg. Therefore there's no actual possibility to retrieve an `f32` value from a `va_list` because of the promotion. This is what I understood too. In that case, there's no possibility of retrieving anything other than a C `int`, `unsigned int`, or `double`... so why the impls for the other types?
> Precisely because those impls are supposed to handle the promotion behavior correctly.

I do not believe this should be done. The Rust compiler does not even correctly check _passing_ varargs currently when generic types become involved, see https://github.com/rust-lang/rust/issues/61275 about that. When combined with accepting these types, this would make it very difficult to reason correctly about the type conversions involved, because we would be adding in our own special rules on top of the already Byzantine rules for C's variable arguments. This would make it much harder to correctly add support for passing e.g. structs over variadic arguments, down the line. I don't believe our current implementations for variadic arguments are adequately or thoroughly tested. We have had a bad habit of not even testing.

In other words, if you extract a different type from the `va_list` than the types that are passed into the variable arguments (barring some **very** specific equivalencies that are preserved between types of the same size), it's UB in C. It should be UB in Rust, too.

I have found way too many bugs in how we handle low-level compilation details, including ABI problems, to believe this is a good idea.

**Comment by workingjubilee** [comment]
It also seems incoherent to provide promotion behavior for all of those types and then refuse to promote f32 (or f16), I would think, as f32 can in fact be correctly translated to f64.

**Comment by workingjubilee** [comment]
@programmerjake has informed me we cannot even rely anymore on things like "all C types will get promoted to at least a certain size" (even relative to the C ABI!) casting further doubt on that idea.

**Comment by programmerjake** [comment]
yes, afaict `_BitInt(8)` doesn't get promoted to `int` but instead gets passed as an 8-bit integer (verified using `clang -emit-llvm`), this also applies to FP types -- `_Float32` doesn't get promoted to `double` (didn't check what clang does). just the older C types get promoted: e.g. `signed char` to `int` and `float` to `double`.

**Comment by RalfJung** [comment]
> In other words, if you extract a different type from the va_list than the types that are passed into the variable arguments (barring some very specific equivalencies that are preserved between types of the same size), it's UB in C. It should be UB in Rust, too.

Definitely fully agreed. Whether an `f32` gets promoted to something else or not should not be the programmer's concern. Especially if that is already how C works (if the argument has type `f32`, it needs to be extracted from the va_list at type `f32`), then we should do the same in Rust.

It follows that the entire discussion about what gets promoted is unnecessary. It's the backends responsibility to handle this correctly.

**Comment by programmerjake** [comment]
> Whether an `f32` gets promoted to something else or not should not be the programmer's concern.

I think Rust shouldn't do promotions for `f32` since there are multiple corresponding C types: `float` and `_Float32`. `float` gets promoted but `_Float32` does not, so I think the programmer just has to manually promote to `f64` if that's what the function expects, Rust doesn't know wether `float` or `_Float32` is what the API expects so can't do promotions for you.

**Comment by RalfJung** [comment]
As an example of the kind of "fun" one can have with this, see https://github.com/rust-lang/rust/issues/71915. The man page says the argument has type `mode_t`, which is `unsigned short`, but the actual function signature is a variadic so on the ABI level the argument is passed as `int`... and if you pass this as a `mode_t` you risk UB.

**Comment by programmerjake** [comment]
maybe a good solution would be to warn when passing or accepting types that are usually subject to promotion since `_BitInt(N)` and `_Float32` are much rarer than `char`, `short`, and `float`. when you actually need to pass/accept `_BitInt(8)` or `_Float32` you can use an attribute to silence the warning, maybe `#[uncommon_c_types]`:
```rust
unsafe extern "C" {
    unsafe fn f(c_int, ...);
}

pub g(v: f32) {
    unsafe {
        f(0, #[uncommon_c_types] v)
    }
}
```

**Comment by lolbinarycat** [comment]
why are you suggesting a special attribute for silencing a warning, we already have `allow`.  such a warning does seem like a good idea though, mixing up `u8` and `char` in signatures does seem like it could be problematic.

**Comment by programmerjake** [comment]
> why are you suggesting a special attribute for silencing a warning, we already have `allow`.

because iirc a lot of people don't want rustc to emit warnings on code when it's the only valid way to implement something and you can't rewrite it to be warning-free.

**Comment by lolbinarycat** [comment]
> because iirc a lot of people don't want rustc to emit warnings on code when it's the only valid way to implement something and you can't rewrite it to be warning-free.

this is still doing that, just providing an alias for `allow(whatever)` doesn't change that. i would argue such a warning should probably be part of clippy instead.  also the attribute feels badly named, since it doesn't mention the important part, is which is that this has to do with *variadics* and *promotion*.

**Comment by workingjubilee** [comment]
> Definitely fully agreed. Whether an f32 gets promoted to something else or not should not be the programmer's concern. Especially if that is already how C works (if the argument has type f32, it needs to be extracted from the va_list at type f32), then we should do the same in Rust.

To be clear, this is actually part of the issue: in C, you *are* expected to have memorized the arcane promotion rules, and thus the author of the variadic function *has* to specify `va_arg` to name exactly the type *that it has been promoted to*. To do otherwise is UB.

This is because the C promotion rules actually _are not_ about the ABI. As demonstrated by `_BitInt(8)`, it is legal, in some cases, to pass smaller-than-`c_int` args. Instead, it has to do with *C source code semantics*: all of these transitions are handled by **clang's frontend**, before it ever reaches LLVM!

So, I think that in Rust, we should not try to squirm out from under this. We should make apparent the bizarreness of the results.

The example you cite, in any case, actually would not compile if you had tried to make the change: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=26aa043833877711d0aba4e26abdebae

**Comment by programmerjake** [comment]
> this is still doing that, just providing an alias for `allow(whatever)` doesn't change that.

I guess...

> i would argue such a warning should probably be part of clippy instead.

I strongly think it should be in rustc because >99.9% of the time the warning is correct and C APIs are almost always documented as you just pass the smaller type, but rust needs the argument to be cast to the promoted type.

rustc already has lints that will be incorrect more often than this, e.g.: [temporary_cstring_as_ptr](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d67907e0bc0614144beece15b4d47b34)

> also the attribute feels badly named, since it doesn't mention the important part, is which is that this has to do with _variadics_ and _promotion_.

yeah, picking a different name is fine with me.

**Comment by RalfJung** [comment]
> in C, you are expected to have memorized the arcane promotion rules, and thus the author of the variadic function has to specify va_arg to name exactly the type that it has been promoted to. To do otherwise is UB.

Fun!

> The example you cite, in any case, actually would not compile if you had tried to make the change: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=26aa043833877711d0aba4e26abdebae

Ah, neat. I had not seen that as I was just working on the other end -- implementing `open` in Miri. The error message itself is not very clear about *why* I need to do this though...
Interestingly, the detailed explanation for that error then says

> Certain Rust types must be cast before passing them to a variadic function, because of arcane ABI rules dictated by the C standard.

which seems to contradict a little your statement that this is not about ABI.

> also the attribute feels badly named, since it doesn't mention the important part, is which is that this has to do with variadics and promotion.

Note that Rust also has a concept called ["promotion"](https://github.com/rust-lang/const-eval/blob/master/promotion.md), and it is entirely unrelated. So we need to be careful with how we use that term in documentation.

**Comment by workingjubilee** [comment]
I don't think the expanded error message is correct, no.

It was written by someone who had far more SAN remaining than me.

**Comment by balt-dev** [comment]
The missing safety contracts on the documentation for VaList really irk me.

**Comment by RalfJung** [comment]
Then maybe make a PR that adds some? :)
Otherwise, your comment isn't going to achieve anything I am afraid. We all know that the feature is quite incomplete.

**Comment by RalfJung** [comment]
On another note, there was some recent discussion of varargs [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/136281-t-opsem/topic/varargs) which should be helpful on writing a spec for the ABI compatibility requirements of this feature.

**Comment by workingjubilee** [comment]
A note re: "argument forwarding". It has been brought up but I do not believe that "argument forwarding" is something we should engage with. If you need `sprintf` then call `vsprintf`. We will become able to declare that `extern "C" fn` correctly with this extension.

**Comment by folkertdev** [comment]
We need `c_variadics` on stable for zlib-rs, so I'm going to try and move this forward. A draft for a modified API is at 

https://gist.github.com/folkertdev/47c79e2f5b03f1138db9d53be5e51ed1

Some highlights

- `VaListImpl` is private!
- `VaList` just has a single lifetime parameter now, which makes a lot more sense
-  The `VaList::with_copy` function and a new `va_copy!` macro can be used to duplicate a `VaList`

This all relies on a different desugaring of `...`, roughly

```rust
fn foo(placeholder: i32, mut args: ...) {
    // stuff
}

// --->

fn foo(placeholder: i32) {
    let tag = core::mem::MaybeUninit::<VaListImpl>::uninit();
    unsafe { core::instrinsics::va_start(tag.as_mut_ptr()) }
    VaList(tag.assume_init_mut())

    // stuff
}
```

It seems possible to make that work internally.

---

I think this new API is nicer, because it leaks fewer implementation details. But before we start implementation work, does anyone see issues with this direction?

pinging specifically the c2rust folks @thedataking @ahomescu, you've probably worked with the existing `c_variadics` most.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

Let's fork discussion of that proposal out to:

- https://github.com/rust-lang/rust/issues/141524

We'll take the nomination there.

**Comment by RalfJung** [comment]
In [#t-compiler > c_variadic API and ABI @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/c_variadic.20API.20and.20ABI/near/527165748), we are puzzling quite a bit about https://github.com/rust-lang/rust/pull/59625 -- a lot of the example and motivation don't make sense to us. If anyone involved in that PR is still around, it'd be great if you could shed some light on the situation. :) Cc @eddyb @ahomescu @dlrobertson

**Comment by ahomescu** [comment]
> In [#t-compiler > c_variadic API and ABI @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/c_variadic.20API.20and.20ABI/near/527165748), we are puzzling quite a bit about [#59625](https://github.com/rust-lang/rust/pull/59625) -- a lot of the example and motivation don't make sense to us. If anyone involved in that PR is still around, it'd be great if you could shed some light on the situation. :) Cc [@eddyb](https://github.com/eddyb) [@ahomescu](https://github.com/ahomescu) [@dlrobertson](https://github.com/dlrobertson)

Catching up to this whole discussion, I'll respond on Zulip.

**Comment by obi1kenobi** [comment]
If this is near stabilization, just mentioning I'd also love to chat about SemVer expectations here. I have some intuition but I'm sure there are edge cases the folks here would catch and I'd miss. I'd love your help!

**Comment by RalfJung** [comment]
> In another thread it was also suggested to rename VaArgSafe to VaArg. I think this fits in with other marker traits like Copy, Send and Sync.

Yeah I saw @workingjubilee propose that. TBH I don't fully get it, `VaArgSafe` seems fine to me. "This type is var-arg" doesn't make a ton of sense -- variables can be varargs, but not types.

Jubilee also proposed not having the trait sealed. That seems odd to me; if we allow user-defined impls we'll have to define how varargs work for every possible type -- avoiding that is a big reason why the trait exists, in my view.

**Comment by programmerjake** [comment]
having `VaArgSafe` require `Self` be either one of the types `std` impls it for or a `#[repr(transparent)]` wrapper over some `impl VaArgSafe` type seems good enough for me -- the compiler could check the impls like it checks `Drop` or `Copy` impls and/or the trait can be `unsafe`. This allows us to not need to define how it works for every type since either we defined it in `std` or it's equivalent to a type in `std` or it's a compile error and/or UB.

later, we can extend that to allow `#[repr(C)]` structs/unions and FFI-safe enums.

**Comment by tgross35** [comment]
Also on the bikeshed, I think it would be nice to have all these types in a new `core::ffi::va` module (possibly with reexports). It makes sense to group the related API, and gives us a more obvious home for va-related documentation.

**Comment by chorman0773** [comment]
As a measure prior to any extension to `VaArgSafe`, is there a reason there can't be a `VaList::arg_unchecked` that also implies the safety condition of not being a promoted type?

**Comment by folkertdev** [comment]
Let's formally ask T-libs-api what they think of the API and names, given that this should be quite close to stabilization. Here is the current state of the API:

```rust
#[repr(transparent)]
#[lang = "va_list"]
pub struct VaList<'a> {
    inner: VaListInner, // target-specific
    _marker: PhantomCovariantLifetime<'a>,
}

// our version of va_copy
impl<'f> const Clone for VaList<'f> { /* ... */ }

// our version of va_end
impl<'f> const Drop for VaList<'f> { /* ... */ }

// What this trait is implemented on depends on the target. Generally it is
// 
// {i, u}{32, 64, size}, f64, *mut T, *const T
// 
// but we also always implement for c_int and c_uint, and c_float, which are 16-bit and 32-bit on avr.
// Types smaller than c_int, c_uint or c_double are promoted in C, and cannot implement VaArgSafe.
// We have a custom error message in place suggesting the appropriate cast.
pub trait VaArgSafe: Copy + sealed::Sealed {} 

impl VaList<'_> {
    // our version of `va_arg`
    unsafe fn next_arg<T: VaArgSafe>() -> T { /* ... */ }
}
```

open questions:

- the name of `VaArgSafe` is unlike most other markers (`Copy`, `Send`, `Sync`), but there is precedent in `UnwindSafe` and `RefUnwindSafe`. Any objections here?
- should the name of `arg` actually be `next` (or some variation like `next_arg`)
- should any other type implement `VaArgSafe`? Technically `char` could work, also the `NonZero` versions of the integers, `NonNull` and references. These types all have additional safety invariants though, and I think it's best to have users handle those explicitly.
- We can support `i128`/`u128` on some targets, but apparently not on all targets (see [#t-lang > Is varargs f32 &#96;float&#96; or &#96;_Float32&#96; @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/Is.20varargs.20f32.20.60float.60.20or.20.60_Float32.60/near/586275135)). The same will likely be true for `f128` eventually. I'm inclined to leave it out of the stabilization for now, and add it behind its own feature gate.  
- should we add implementations for `MaybeUninit<T>` (for the `T`s that currently implement the trait)-
- is exposing from `core::ffi::VaList` OK? should this get its own module to export from?

NOTE: this is a summary for T-libs-api, for discussion on these topics we should probably use [#t-libs](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs) rather than github comments here.

**Comment by folkertdev** [comment]
Discussed today ([notes](https://hackmd.io/d9D6vUnuTnCWygkc3hffEw#nominated-rusttf44930-Tracking-issue-for-RFC-2137-Support-defining-C-compatible-variadic-functions-in-Rust-c_variadic)), we concluded:

- `VaArgSafe` is fine as a name
- we want to rename `arg` to `next_arg`; `next_unchecked` was discussed, but rejected because there will never be a checked version
- we agreed to only implement `VaArgSafe` for types that can be defined in C (so `c_int`, `c_double`, etc.), and not do it for `char` or `NonZeroI8` or even `Option<NonNull>`. Users can perform that conversion themselves
- because whether `i128` is FFI-safe is platform-specific, we should implement `VaArgSafe` only on targets where the type can be mapped to `__int128`. We'll add that impl for the main targets now, and can extend support in the future as needed
- we defer implementing on `MaybeUninit<T>` until an actual use case comes up. It's a bit weird what this type corresponds to on the C side
- because there are only 2 items (the type and the trait) we decided to just leave them in `core::ffi` and not give them their own publicly-visible module and just document the struct (with some general context on what c-variadic functions are/do) and the trait (with the rational for why only some types implement that trait)

**Comment by folkertdev** [comment]
A (hopefully) quick question for T-lang

https://github.com/rust-lang/rust/blob/cf1817bc6ecd2d14ca492247c804bad31948dd56/compiler/rustc_codegen_llvm/src/va_arg.rs#L1178-L1193

There are some targets where we're not 100% confident that the implementation is correct. 

- mips
- 32-bit sparc
- m64k and msp430
- custom targets that use a custom target json file with an unrecognized architecture

There are no known issues, and to our knowledge we match clang in all cases, but these targets are niche and hard to run, and the target maintainers have not responded so far. How should we handle this?

- just do what clang does (i.e. defer to LLVM's `va_arg`) and hope for the best. these are all tier-3 targets, so shrug 
- use `bug!` to trigger an ICE on those targets
- gate the implementation behind a feature, e.g. `c_variadic_experimental_arch`

**Comment by RalfJung** [comment]
I think I'd prefer `bug!` (or a nicer error). We don't have a clear "tier 2 target checklist" (we probably should) so for all we know these targets will be made tier 2 without anyone ever properly testing this. Seems better to error saying "variadics not supported on this target, please file a bug if you can help us test them".

**Comment by folkertdev** [comment]
I'd personally prefer the feature because at least it gives users of these targets a workaround (it seems fair to assume they are familiar and comfortable with using nightly), and we'd hopefully get feedback if there do turn out to be issues.

Also when a target would want to upgrade to tier 2 we'd add support behind a feature gate anyway (presumably), so we may as well just do that from the start.

**Comment by joshtriplett** [comment]
Dropping lang nomination, as the open question got answered in today's lang meeting, resulting in https://github.com/rust-lang/rust/pull/155974 .

**Comment by joshtriplett** [comment]
There's a stabilization PR and stabilization report in https://github.com/rust-lang/rust/pull/155697 . I've started a lang FCP on that PR. https://github.com/rust-lang/rust/pull/155697#issuecomment-4348080495 discusses the current aspects in flight.

The libs-api portion of this also needs stabilizing at the same time. As libs-api does FCPs on tracking issues instead, I'm starting a libs-api FCP on this tracking issue. libs-api, shall we stabilize the `VaArgSafe` sealed trait and `VaList` type?

@rfcbot merge libs-api

**Comment by rust-rfcbot** [comment]
Team member @joshtriplett has proposed to merge this. The next step is review by the rest of the tagged team members:

* [x] @Amanieu
* [ ] @BurntSushi
* [x] @dtolnay
* [x] @joshtriplett
* [ ] @the8472

No concerns currently listed.

Once a majority of reviewers approve (and at most 2 approvals are outstanding), this will enter its final comment period. If you spot a major issue that hasn't been raised at any point in this process, please speak up!

cc @rust-lang/lang-advisors: FCP proposed for lang, please feel free to register concerns.
See [this document](https://github.com/rust-lang/rfcbot-rs/blob/master/README.md) for info about what commands tagged team members can give me.

**Comment by rust-rfcbot** [comment]
:bell: **This is now entering its final comment period**, as per the [review above](https://github.com/rust-lang/rust/issues/44930#issuecomment-4348137769). :bell:

**Comment by rust-rfcbot** [comment]
The final comment period, with a disposition to **merge**, as per the [review above](https://github.com/rust-lang/rust/issues/44930#issuecomment-4348137769), is now **complete**.

As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.
